types/src/transaction/mod.rs

```diff
@@ -538,6 +762,158 @@ impl TransactionPayload {
             payload => panic!("Expected EntryFunction(_) payload, found: {:#?}", payload),
         }
     }
+
+    pub fn has_nonce(&self) -> bool {
+        match self {
+            Self::Payload(TransactionPayloadInner::V1 {
+                extra_config:
+                    TransactionExtraConfig::V1 {
+                        replay_protection_nonce,
+                        ..
+                    },
+                ..
+            }) => replay_protection_nonce.is_some(),
+            _ => false,
+        }
+    }
+
+    pub fn executable(&self) -> TransactionExecutable {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                TransactionExecutable::EntryFunction(entry_function.clone())
+            },
+            TransactionPayload::Script(script) => TransactionExecutable::Script(script.clone()),
+            TransactionPayload::Multisig(multisig) => multisig.as_transaction_executable(),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                executable.clone()
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                unimplemented!("ModuleBundle variant is deprecated")
+            },
+        }
+    }
+
+    pub fn executable_ref(&self) -> Result<TransactionExecutableRef> {
```

Comment:
oh interesting now I understand why we need this one, curious how much performance difference it is

Reply:
Haven't measured performance difference. This avoids a few clone operations in aptos_vm.rs. As cloning payload could be expensive, seemed worth it.